### DISCUSSION TITLE
Set bluetooth mac address on ESP32?

### DISCUSSION THREAD

**Discussion by harbaum** [micropython/micropython#10080] [Libraries & Drivers]
I am trying to act like a fischertechnik toy controller. The matching other device will only connect to devices of a specific vendor, in this case LNT (vendor prefix 10:45:F8).

Can I change the BT MAC address through micropython? I tried:

`bluetooth.BLE().config(mac=(0, b'\x01\x02\x03\x04\x05\x06'))`

But that results in:

`ValueError: unknown config param`

**Comment by puppet13th** [comment]
i think it's not possible to set specific bluetooth mac address. check [doc](https://docs.micropython.org/en/v1.15/library/ubluetooth.html) about 'addr_mode'

**Comment by harbaum** [comment]
It seems the esp-idf only recently got the API to change individual mac addresses:

https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/misc_system_api.html#mac-address
`
esp_err_t esp_iface_mac_addr_set(const uint8_t *mac, esp_mac_type_t type)`

It's not even in  the 5.0 release. It seems there were ways to change single or all mac on previous versions as well but this seems to be the first time there's a clean API to do this.

**Reply by puppet13th** [reply]
If custom mac address is supported in IDF, there is need to be implemented in mpy function. @dpgeorge @jimmo any comment on this?

**Reply by harbaum** [reply]
I really don't know how important this use case is. On the other hand many MPY/esp32 are probably used to emulate existing devices. And this may come with the need to change mac addresses.

**Comment by harbaum** [comment]
I just tested to add support for esp_base_mac_addr_set() to MPY and it works just fine. Downside of this approach is that it needs to be done before any network related code is run and it affects all mac addresses (Wifi station, wifi ap, bluetooth and ethernet).

I'll append a patch here.

**Reply by shr00mie** [reply]
Looks like there's more than one way to skin this particular cat:

<img width="711" alt="image" src="https://user-images.githubusercontent.com/7096317/229384910-9facbd58-8d1f-4f86-ad3a-aa5ff3dcc27d.png">

* old api
  * esp_base_mac_addr_set()
    * sets the base mac address (for the session, from the looks of it), from which the other interface MACs are derived
* new api
  * esp_iface_mac_addr_set()
    * replacement for old API. used in conjunction with ESP_MAC_BASE argument to set the base MAC for the device (again, seems like for the duration of the session).
  * [esp_netif_set_mac()](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_netif.html#_CPPv417esp_netif_set_macP11esp_netif_tA_7uint8_t)
    * this seems to be the underlying function we would want exposed when calling IFACE.config(mac=''), especially because:
    "It is also possible to call the function esp_netif_set_mac() to set the specific MAC used by a network interface after network initialization."
    * ~~better yet, introduce logic that checks if the interface in question, or any interface for that matter, has been initialized, and call esp_iface_mac_addr_set() with ESP_MAC_BASE if no interfaces have yet to be initialized, else fall back to esp_netif_set_mac(), which, at least per espressif's documentation, can be called even if the interface is already up.~~
    * i take that back.
      * seems like esp_iface_mac_addr_set() should be used when you want to, for example, maintain a certain manufacturer prefix, but don't care what the incremented interface addresses might be
        * this should probably be set somewhere in boot.py and be implemented under machine, as you have already done.
      * after initializing a given interface, if the config method is called with the mac argument, expose esp_netif_set_mac() should be exposed underneath, allowing those MACs to be set regardless of whether the associated interface has been brought up or not.

it's also entirely possible i have no idea what i'm saying.

[LiNkY](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/misc_system_api.html#mac-address)

p.s. it just seems like esp_iface_mac_addr_set() has the potential to be considerably more problematic due to, at least based on my reading, it setting the base device MAC, instead of the desired interface MAC, due to:

![image](https://user-images.githubusercontent.com/7096317/229398462-0d50aa2e-e7c2-404c-9506-ce58c75eaa8a.png)

i suppose it's all well and good if you only ever plan on setting a custom mac for the esp32 as a wifi client, as that just happens to mirror the burnt in and/or custom base MAC, but for the remaining interfaces, it would seem that one would have to take the desired interfaces' last octet offset into consideration with additional code to offset the offset, which sounds about as appealing as jumping head first into a dumpster fire.

functionality has been around for a while. doesn't show up if you just change the esp-idf version via the dropdown. gotta go dig around for it. being able to set the hardware address on an interface seems like something fundamental, and has probably been around considerably longer than v4.4.

[custom mac esp-idf v4.4](https://docs.espressif.com/projects/esp-idf/en/release-v4.4/esp32/api-reference/system/system.html#mac-address)

**Comment by harbaum** [comment]
So here's the little patch that allows to read and write the base mac address on esp32. 

[modmachine_mac_addr.patch.txt](https://github.com/micropython/micropython/files/10098686/modmachine_mac_addr.patch.txt)

Initially I hoped that I could use this to first read the base mac, then modify it and write it back. Turned out that you cannot read the base MAC before you set it. So you cannot use that to read the base mac set by the system itself. But unique_id() returns exactly that. So adding this to boot.py will move the mac address into the LNT range:

```
# modify mac for LNT vendor 10:45:F8 
mac = machine.unique_id()
machine.base_mac_addr(b'\x10\x45\xf8' + mac[3:])
```

Afterwards the device shows up using the 10:45:f8 vendor prefix for wifi as well as bluetooth.

**Reply by qaq112233** [reply]
How to use this patch?

**Reply by luxlux** [reply]
Hi Til,

i want to clone a BLE device, to find out which commands the manufacturer app sends to it.

So after finding your post, i spend a full day/night, to get my first build ever of an esp firmware done. i hope i have applied your patch correctly in my fork: https://github.com/luxlux/micropython/blob/master/extmod/modmachine.c (from line 143 and 273)

But now, if i try a
`from machine import base_mac_addr
`
in Thonny i get only an "`ImportError: can't import name base_mac_addr`" back.

Was possibly your patch not  complete? 
I never did C development. do possibly the .h file need also to be patched? 
If yes: how?

hope you can still help me after all the time. (i also searched your micropython rep for your applied patch, but found nothing)...

**Reply by biemster** [reply]
I've used this patch successfully a while ago, and have a MicroPython for esp32-c3 and one for s3 somewhere with this, what esp32 are you on?

**Reply by luxlux** [reply]
Some hours later i also got it.
I needed to change the Word "STATIC" to "static" and needed to remove also one static. Also i added a line at modmachine.h:
`MP_DECLARE_CONST_FUN_OBJ_VAR_BETWEEN(machine_base_mac_addr_obj);
`
I don't know if the change in modmachine.h was really needed. But it compiled successfully and now i can set the MAC address!

I also always use the cheap esp32-c3 bords: https://de.aliexpress.com/item/1005006283171104.html
Today i got also this one, which also worked with the same build: https://de.aliexpress.com/item/1005007342383107.html

And i was successful! I was able to receive commands from the app of the electronic scale manufacturer, with my fake scale board using the same MAC address. I simply connected the original scale to the app, then i took out the batteries, and started the well prepared micropyton board. Then i pressed some keys in the app. Now i know which data i have to send to the scale to remotely press the "TARE" key, or change the weighting unit:
https://juvenile-tortoiseshell-ptarmigan.glitch.me/index-scale.html

**Reply by Temgle** [reply]
Hi luxlux:
Could you share this firmware for ESP32?

**Reply by biemster** [reply]
@Temgle While @luxlux is finding a way to upload, there is a 6 month old C3 firmware with the patch here: https://github.com/biemster/FindMy/tree/main/ESP32

**Reply by luxlux** [reply]
here it is. It is NOT for ESP32!! It is for the cheaper ESP32-**C3**, which uses a only 1 core and a different architecture, than the dual core ESP32...
https://github.com/luxlux/micropython/releases/tag/v0.1-alpha

**Reply by Temgle** [reply]
Hi Luxlux, thank you for your help. I try it on ESP32-C3, but it didn't work:
![image](https://github.com/user-attachments/assets/18db13b1-83cf-4567-a56c-ba4b406c1cf6)

**Reply by luxlux** [reply]
Which Computer OS do you use? And which ESP32C3 board?

Have you did these two steps:

First: erase your board:
`esptool.py --chip esp32c3 --port /dev/tty.usbmodem1101 erase_flash`  (instead of "/dev/tty.usbmodem1101" you need to use the correct part name to the device, depending on your system)

Second: flash all 3 files at the correct position on your board with:
`python -m esptool --chip esp32c3 -b 460800 --before default_reset --after hard_reset write_flash --flash_mode dio --flash_size 4MB --flash_freq 80m 0x0 bootloader.bin 0x8000 partition-table.bin 0x10000 micropython.bin` (exaclty this command worked for me with 3 different ESP32-C3 boards. I have flashed them from mac os).

Your screenshot shows an other ESP-ROM "ESP32c3-api1-20210207". this is not the rom i have build. my build is from 2025 not 2021.
It looks like you do something wrong, when trying to flash it.
Do you have esptool.py installed on your system?

**Reply by Temgle** [reply]
Thanks Luxlux. I solve the problem by compling myself using your {modmachine.c and modmachine.h} to replace that in the source code.
I use WSL，ESP-IDF version is v5.2.2. And I use thonny to flash it.
But I find the **first byte** of the mac address cannot be set freely, some numbers will have ValueError code: Changing base MAC addr failed.
It's likely that the first byte even number will be OK，odd number will be not.
![image](https://github.com/user-attachments/assets/4a31fd9d-79f7-4127-a808-95899bce87f5)

**Reply by luxlux** [reply]
I don't know. For me with mac 0x3C, 0x27, 0x6F, 0xF0, 0x91, 0xE7 it worked. This was the one which i needed.  I have not tried other mac address. Any how, the setting of the mac address will be executed by EPS-IDF. Possibly you also try a newer IDF version than 5.2.2. I used 5.4.0, there is also a still newer version. If it was an error, this possibly helps. But may be it is limited by design, like by some rules for allowed mac addresses...?! ... You could check the code of IDF if limits are implemented there...

**Reply by luxlux** [reply]
I have now also tried your 2 mac addresses: 
0xFD, 0x92, 0x3A, 0xC1, 0xDA, 0xCA - fails
0xCA, 0xCC, 0x00, 0xFF, 0xFF ,0xFF - works

**Reply by Temgle** [reply]
I found it in the docs of ESP-IDF：Base MAC must be a unicast MAC (least significant bit of first byte must be zero).
![image](https://github.com/user-attachments/assets/774d0335-6ecc-4131-885e-78496e925255)

**Reply by luxlux** [reply]
I have checked [wikipedia](https://en.wikipedia.org/wiki/MAC_address). Your not working address falls into a special range of mac addresses "Multicast (group)". Possibly ESD-IDF does not support Multicast, and so such addresses are not allowed ...

**Reply by Temgle** [reply]
I find a way to set the first byte of mac address without restriction: 
comment out the code line 152-155 in (esp-idf\components\esp_hw_support\mac_addr.c)
![image](https://github.com/user-attachments/assets/c11592ca-65aa-46da-8fb6-3443aa7fde87)

**Reply by luxlux** [reply]
So possibly this is a way to go. ... but keep in mind, this limiting rule could possibly make sense, because i guess also other things have to be fulfilled from ESP-IDF, for a multicast MAC address, which are possibly not fulfilled in the IDF code, so that they disallowed the use of such multicast MAC addresses as Base Mac Address.

**Reply by ctschach** [reply]
This is awesome....can someone create a full, updated patch for the current MicroPython version.  Maybe even a pull request so that this goes into the official release?!

**Reply by luxlux** [reply]
i would say no. this patch works. but it is not really consequent to put it into "machine". for a patched solution this ok. But for a pull to the offical release, it would make more sense, to have realize this it in "bluetooth" or in "aioble".

**Reply by shr00mie** [reply]
Ideally, functionality should be exposed for interfaces to mirror espressif source. This hacky patching of machine address by offset to get the desired ble address is just silly. For some IOT functionality, being able to take advantage of address space in a direct manner is critical (beacons for example).

**Comment by shr00mie** [comment]
alright. so. i think i did a thing. i mean it's ghetto and i'd really love for this to be implemented the way it should be, at the BLE level, but this should work in the meantime.

per the documentation, given the base device MAC, the last octet of the BLE MAC is offset by 2 bits.

which presents itself as an issue if you try to set your desired BLE MAC using the base_mac_addr method:

```python
from machine import base_mac_addr
import bluetooth as bt

# let's say this is the desired BLE mac
mac = bytearray([0x1C, 0x88, 0x3F, 0xE5, 0x7E, 0x8D])

# set device base mac addr
base_mac_addr(mac)

ble = bt.BLE()
ble.active(True)
ble.config('mac')
(0, b'\x1c\x88?\xe5~\x8f')
```

![image](https://y.yarn.co/71551909-bfc5-4a7c-92c9-cc38fa69b602_text.gif)

which can be programmed around like so:

```python
from machine import base_mac_addr
import bluetooth as bt

mac = bytearray([0x1C, 0x88, 0x3F, 0xE5, 0x7E, 0x8D])

def base_ble_offset(mac):
	mac[5] = mac[5] - 2
	return mac

base_mac_addr(base_ble_offset(mac))

ble = bt.BLE()
ble.active(True)
ble.config('mac')
(0, b'\x1c\x88?\xe5~\x8d')
```

![image](https://media.tenor.com/NWqisN5ga_MAAAAC/voila-iron-man.gif)

did i do it right?

**Reply by harbaum** [reply]
This offset is by purpose. The ESP32 needs several MACs for WiFi, Bluetooth etc and it derives them from the base address by just counting up.

**Reply by shr00mie** [reply]
...yeah. i get that part. the point i was trying to make was that it would be nice to be able to set the BT MAC directly via Python, similarly to how the capability is exposed with WiFi, instead of having to set the base address, with additional code in order to account for the offset.

**Comment by ctschach** [comment]
Is there a chance this patch will make it into the official MicroPython distribution?!
